validation: ensure assumevalid is always used during reindex

Eunovo opened the pull request:
assumevalid may be ignored during reindex for the following reasons:
- the assumevalid block is not in the block index:  this happens when the assumevalid block could not be downloaded before the previous IBD run was interrupted
-  the chainwork of the best header is less than the minimumchainwork: happens if the previous IBD was interrupted before it could connect blocks up to minimumchainwork. See Issue https://github.com/bitcoin/bitcoin/issues/31494

This PR fixes this by skipping script verification during reindex if assumevalid is enabled. 

EDIT: While the minimumchainwork is typically set to match the assumevalid block height, users typically adjust the assumevalid block without also adjusting the minimumchainwork; Scenario 2 from the list can occur if the assumevalid block is set to an older block without also reducing the minimumchainwork.

Eunovo committed:
test: Check that reindex always uses -assumevalid

Also move feature_reindex.py test to "Tests less than 2m" section in test_runner.py,
because it is now much slower and has similar runtimes to other tests in this section

Eunovo committed:
validation.cpp: use assumevalid during reindex

-reindex deletes the block tree db and attempts to rebuild it based on local block files,
without attempting to re-request headers for which we don't have the full block.
This might cause:
- the best known chain to have a chainwork less than the minimumchainwork
- the assumedvalid block to not be in the block index if the previous IBD was interrupted
before downloading it

Eunovo commented on src/validation.cpp:
```diff
@@ -2499,7 +2499,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
             if (it->second.GetAncestor(pindex->nHeight) == pindex &&
                 m_chainman.m_best_header->GetAncestor(pindex->nHeight) == pindex &&
-                m_chainman.m_best_header->nChainWork >= m_chainman.MinimumChainWork()) {
+                (m_chainman.m_best_header->nChainWork >= m_chainman.MinimumChainWork() || m_chainman.m_blockman.LoadingBlocks())) {
```
That's true. I thought https://github.com/bitcoin/bitcoin/issues/31494 was focused on the best header chainwork not being up to minimumchainwork. I could change this to always skip script checks during a reindex, but IIUC having a block on disk doesn't mean that it was connected during the previous IBD. Might be worth it to sync headers first then connect blocks even during reindex.

mzumsande replied:
I really wasn't sure what the best solution is when I opened the issue and still am not, but these are my thoughts:

Yes, there are two reasons why we might not apply assumevalid during a reindex currently:
1.) The best header has not enough chainwork
2.) The assumevalid block is not in our block index

> I could change this to always skip script checks during a reindex, but IIUC having a block on disk doesn't mean that it was connected during the previous IBD. 

I think the question is rather how well we can avoid to connect blocks skipping the script checks during the reindex that we would have fully validated in the original run if we hadn't crashed. Assuming that the user provides the same `-assumevalid`  arg to the reindex run that they did in the original run I can see the following cases:
1.) The user provides an existing block (or nothing, so that the default hash from chainparams is used) that we just haven't downloaded yet because it's later in the best chain - if we don't have it in our index we'd be justified to skip script checks - if we do have it in our index, no special rules are necessary for reindexing. I think that's the most common case for users.
2.) The user provides `-assumevalid=0`  - we should respect that also in the reindex run and fully validate everything.
3.) The user provides a block hash with `-assumevalid` that is non-existing / not in the best chain. I think that's the critical case because here we would have fully validated everything in the original run, but won't validate anything during the reindex. But arguably that's not a common case / the user's fault for providing this bad block, so maybe it can be tolerated?

> Might be worth it to sync headers first then connect blocks even during reindex.

Yes, that's the second option I thought of. I think we can't just stop and wait for the headers to be downloaded - reindex should be able to finish without having a p2p connection, even if that means no assumevalid. But I was thinking of maybe just disabling the relevant `LoadingBlocks()` guards in net_processing, so that we'd download headers in parallel to `ActivateBestChain()`, and would change to use assumevalid after a few thousand blocks or so. That seem a bit messy (we'd still connect _some_ blocks without assumevalid) and requires testing but might work well in practice?!

Eunovo replied:
I created another test to check if blocks rejected in the normal IBD run get loaded during a `-reindex`. The test reveals that it does load the invalid block and it will connect it if script checks are skipped, see https://github.com/Eunovo/bitcoin/commit/1ff4e5ee78418a30ba1ea9c384b8c7ee43435ce6

IIUC During a reindex, the node does not know what blocks were connected in the previous run so it can't safely skip checks. It looks to me like we have to:
1. Sync headers before connecting blocks during  reindex OR
2. Persist data to show that we previously connected this block (IIUC this exact data exists but gets deleted during reindex, maybe we can preserve the relevant data we need?)

Eunovo replied:
> The test reveals that it does load the invalid block and it will connect it if script checks are skipped, see [Eunovo@1ff4e5e](https://github.com/Eunovo/bitcoin/commit/1ff4e5ee78418a30ba1ea9c384b8c7ee43435ce6)

I'm still not sure if this causes any damage because I expect the node to eventually pick the most-work chain when it syncs headers

Eunovo commented:
I had an offline conversation with @mzumsande. During the reindex, we can apply assumevalid for following cases:
- If the assumevalid block is the index, then we can skip script checks for all ancestors of the assumevalid block
- If the assumevalid block is not in the index, then the initial IBD run had not downloaded the assumevalid block yet then the blocks on disk are ancestors of the assumevalid block and we can still skip script verification

Then we download headers from peers in parallel to ActivateBestChain()

I'll put this in draft while I implement and test this over  the next few days

l0rinc commented:
I'm not sure whether `LoadingBlocks` is the correct way to solve it or not (I would need someone else to opine on that, I can't tell if this can err on the side of not checking the scripts at all in case of some edge case).

I can confirm however that this does seem to solve the issue I had, reported in https://github.com/bitcoin/bitcoin/issues/31494 (i.e. reindex(-chainstate) after a partial IBD always enabling script verification).

I have added a `LogWarning("fScriptChecks is %s", fScriptChecks);` after the assignments and tested it with master vs the branch:
```bash
   cmake -B build -DCMAKE_BUILD_TYPE=Debug && cmake --build build -j$(nproc) \
&& mkdir -p demo && build/src/bitcoind -datadir=demo -stopatheight=100 \
&& build/src/bitcoind -datadir=demo -reindex-chainstate -stopatheight=10
```

> master
```bash
[warning] fScriptChecks is true
```

> Eunovo:31494-reindex-assumevalid
```bash
[warning] fScriptChecks is false
```

----

I can also confirm that adding `-assumevalid=0` will enable script checking from the beginning and that `-assumevalid=000000009b7262315dbf071787ad3656097b892abffd1f95a1a022f896f533fc` has 5 fScriptChecks=false values after which script checking is (correctly) enabled again.

---

I also checked that the new `assume_valid` test passes here and fails correctly on master (though it's *very* slow, we have to check if there's a way to speed it up somewhat)

<details>
<summary>Details</summary>

```bash
2025-01-30T15:08:22.897000Z TestFramework (INFO): Success
2025-01-30T15:08:31.343000Z TestFramework (INFO): Restarting node while reindexing..
2025-01-30T15:08:32.551000Z TestFramework (INFO): Testing reindex with -assumevalid
2025-01-30T15:09:23.963000Z TestFramework (INFO): Testing reindex with assumevalid block in block index and minchainwork > best_header chainwork
2025-01-30T15:09:52.592000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/Users/lorinc/IdeaProjects/bitcoin/test/functional/test_framework/test_framework.py", line 135, in main
    self.run_test()
    ~~~~~~~~~~~~~^^
  File "/Users/lorinc/IdeaProjects/bitcoin/build/test/functional/feature_reindex.py", line 203, in run_test
    self.assume_valid()
    ~~~~~~~~~~~~~~~~~^^
```

</details>

----

So from me it's an approach ACK, but someone else needs to validate the implementation details and the corner cases that can result from this.

Eunovo commented:
> I'm not sure whether `LoadingBlocks` is the correct way to solve it or not (I would need someone else to opine on that, I can't tell if this can err on the side of not checking the scripts at all for some edge case).

I agree too. `LoadingBlocks()` is `true` when blocks are being reindexed or  when new blocks are being loaded with `-loadblocks` .  I intend to move away from `LoadingBlocks()` to something that's only true when blockfiles are being reindexed.

> I also checked that the new `assume_valid` test passes here and fails correctly on master (though it's _very_ slow, we have to check if there's a way to speed it up somewhat)

I'll look into this!

mzumsande reviewed (commented):
Concept ACK

I think the current approach would work.
Even though it means that if the user provided a bad / non-existing assumeutxo hash they could skip script validation during a reindex when they wouldn't have in the original run,  I'd  say that's not a major problem because a reindex should only be necessary in case of a a db corruption, which should not be remotely triggerable.

So it will hopefully make the reindex a bit faster in case of a database corruption, while not changing the assumevalid security assumption very much, and it does respect `-assumevalid=0`.

Though it would be good to have some experienced contributors chime in that were around when assumevalid was introduced (#9484).

Eunovo commented:
Thanks for the reviews @l0rinc and @mzumsande . Due to your feedback, I have opened the PR for general review.

mzumsande commented on test/functional/feature_reindex.py:
```diff
@@ -97,6 +121,77 @@ def continue_reindex_after_shutdown(self):
+        blocks.append(invalid_block)
+
+        # Bury that block with roughly two weeks' worth of blocks (2100 blocks) and an extra 100 blocks
+        for _ in range(2200):
```
is there a reason we need to mine that many blocks, and not just a few more to make the test faster?

Eunovo replied:
Are you referring to the 100 blocks mined after the first 2100 blocks?

mzumsande replied:
I meant the 2100 (two weeks) of of blocks. I didn't test this, but is there a reason this couldn't just be a few blocks (plus a few more that aren't sent to the node)?

Eunovo replied:
Script checks are only skipped when the current block is roughly 2 weeks away from the tip. See https://github.com/bitcoin/bitcoin/blob/3f83c744ac28b700090e15b5dda2260724a56f49/src/validation.cpp#L2476-L2494

We have to bury the test block 2 weeks deep for script verification to the skipped

mzumsande replied:
oh I see, forgot about that check - can be resolved.

Eunovo replied:
@stratospher the test completes much faster on my system, but by checking the timing of other tests, I put it among "Tests less than 2m" section because it completes faster than `mining_getblocktemplate_longpoll.py` than slower than `mempool_persist.py`

Eunovo commented on test/functional/feature_reindex.py:
```diff
@@ -97,6 +121,77 @@ def continue_reindex_after_shutdown(self):
+        assert_equal(node.getblock(node.getbestblockhash())['height'], expected_height)
+        assert_equal(int(node.getbestblockhash(), 16), blocks[2100].sha256)
+
+        self.log.info("Success")
```
Will do

Eunovo replied:
Done

Eunovo commented on src/node/blockstorage.cpp:
```diff
@@ -1213,6 +1228,11 @@ void ImportBlocks(ChainstateManager& chainman, std::span<const fs::path> import_
             nFile++;
         }
+
+        // Call ActivateBestChain here so we can skip script verification
```
Done

stratospher reviewed (commented):
ACK 11a9d55e.

Agree that applying reindex-assumevalid combo can be useful in the special cases mentioned above to speedup reindexing when we don't have all blocks.

A behaviour change during assumevalid+reindex now would be:
- we'd skip script checks even when the assumevalid blockindex doesn't exist (ex:we put `assumevalid=somegarbage(or)blockhash-with-typo`)
- this was not the case on master where we skip script checks only when know that the assumevalid blockindex exists.

Don't think it's a problem since assumevalid already puts responsibility on the user and we'd need to have tampered block files on disk from some external source (in which case user shouldn't be using assumevalid in the first place). just wanted to mention it + curious about what others think about this/any other possible behaviour change.

Eunovo commented:
Rebased https://github.com/bitcoin/bitcoin/pull/31615/commits/11a9d55e84af2d6ff138b99310720667cd59dcc3 to https://github.com/bitcoin/bitcoin/pull/31615/commits/e260e23d88cd6b2aaa03d9d45dae8b174f2cfe1f

- moved `feature_reindex.py` test to "Tests less than 2m" section in `test_runner.py`

Eunovo commented:
Rebased https://github.com/bitcoin/bitcoin/commit/e260e23d88cd6b2aaa03d9d45dae8b174f2cfe1f to https://github.com/bitcoin/bitcoin/commit/a53d43bc0f010513b0922ad48d611f402ec0e511 to address https://github.com/bitcoin/bitcoin/pull/31615#discussion_r2185605623

Eunovo commented:
Rebased on master@ https://github.com/bitcoin/bitcoin/commit/5878f35446

l0rinc commented on test/functional/test_runner.py:
```diff
@@ -103,6 +103,7 @@
     'p2p_segwit.py',
     'feature_maxuploadtarget.py',
     'feature_assumeutxo.py',
+    'feature_reindex.py',
```
What's the reason for moving this? If it's needed, could you add it to the commit message?

Eunovo replied:
Done.

l0rinc reviewed (approved):
ACK 01e372b25855a21d205c6ded83f6849691111d42

Thanks for fixing it - someone with more knowledge in this area should also validate it.

Eunovo commented on src/node/blockstorage.cpp:
```diff
@@ -1263,18 +1283,7 @@ void ImportBlocks(ChainstateManager& chainman, std::span<const fs::path> import_
-        }
-    }
+    activate_all_chainstate();
     // End scope of ImportingNow
```
Yes, it does. The ImportingNow object is destroyed at the end of the function.

Eunovo commented on src/node/blockstorage.cpp:
```diff
@@ -1241,6 +1256,11 @@ void ImportBlocks(ChainstateManager& chainman, std::span<const fs::path> import_
+
+        // Call ActivateBestChain before setting m_blockfiles_indexed so we can skip script verification
+        // during reindex if assumevalid is enabled
+        activate_all_chainstate();
```
`ImportBlocks` first tries to load blocks from existing block files during the reindex process. After the reindex process, it tries to load any external blocks provided by the user. We only want to force the use of assumevalid(if set) during the reindex process, not when loading external blocks.

We call `activate_all_chainstate` the first time, during the reindex process. This allows us to skip script checks using modified assumevalid rules for reindex.
We call `activate_all_chainstate` the second time, after external blocks have been loaded, so we can search for better chains among the new blocks. If no external blocks were loaded `activate_all_chainstate` will have no effect.

I have added a comment explaining that the second call will have no effect no external blocks were imported.

I am not convinced that this should be in a separate commit. This change only makes sense in the context of this commit.

l0rinc reviewed (approved):
ACK 89b5b607e3380ff2cf03d8199c70e655e8c265cb

Consider unifying the way your refer to the first block that fulfills the assumevalid condition

Eunovo commented on src/validation.cpp:
```diff
@@ -2486,8 +2486,14 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
             }
+        } else if (m_chainman.m_blockman.IsReindexing()) {
+            // During a reindex, the assumed valid block may not be in the index
+            // if the previous IBD run was interrupted before it downloaded the assume valid block.
```
Done

Eunovo commented on src/node/blockstorage.cpp:
Done

l0rinc commented:
Just tried this on my rpi4b server which keeps enabling signature validation after a reindex. Restarting it with only the args `./build/bin/bitcoind -dbcache=5000 -datadir=$DATA_DIR -assumevalid=0000000000000000000087564caa77e7b3f29d0464256c04d5539e43663f8698` still shows heavy signature validation (based on `perf top`).
<img width="1000"  alt="image" src="https://github.com/user-attachments/assets/b29b6bdd-6be2-4b05-98e3-82d60989c52c" />

Eunovo commented:
Rebased on master and added a comment explaining why we bother to skip minimumchainwork, even though we expect minimumchainwork to match assumevalid block height most times.

l0rinc commented on src/validation.cpp:
```diff
@@ -2449,8 +2449,17 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
                 fScriptChecks = (GetBlockProofEquivalentTime(*m_chainman.m_best_header, *pindex, *m_chainman.m_best_header, params.GetConsensus()) <= 60 * 60 * 24 * 7 * 2);
             }
+        } else if (m_chainman.m_blockman.IsReindexing()) {
+            // During a reindex, the assumed valid block may not be in the index
```
While the change does seem to fix my personal experience of unexpected script validations, I don't yet understand why reindex does this in the first place - can we fix that instead of adding extra conditions to the script validation checks? Seems a bit hacky and only solves part of the problem, doesn't it?

l0rinc replied:
Not necessarily, but it bothers me that we (or at least I) don't understand what's causing the problem in the first place, the current solution feels like a work-around to me - which is still better than what we had, but it's also arguably more complicated and it doesn't seem to fix the underlying problem of why reindex is missing headers in the first place. And are we sure it's a reindex-only bug or can it happen for reindex-chainstate or IBD as well? I haven't dug too deep into this, but I want it to be fixed and was wondering if the apparent lack of interest in the reviews is because others also felt it's a workaround. Thanks for working on this.

Eunovo commented on src/node/blockstorage.cpp:
```diff
@@ -1263,18 +1283,10 @@ void ImportBlocks(ChainstateManager& chainman, std::span<const fs::path> import_
-        }
-    }
+    // Call activate_all_chainstate() again, after loading blocks from files,
+    // If no blocks were imported, ActivateBestChain will have nothing to do
```
Thanks for pointing this out. I will update the comment.

hodlinator commented on test/functional/feature_reindex.py:
```diff
@@ -97,6 +121,90 @@ def continue_reindex_after_shutdown(self):
+        assert_equal(node.getblock(best_blockhash)['height'], block_info['height'])
+        assert_equal(best_blockhash, block_info['hash'])
+
+        self.log.info("Success")
```
meganit:
```
₿ git grep 'self.log.info("Success")'
test/functional/feature_pruning.py:        self.log.info("Success")
test/functional/feature_pruning.py:        self.log.info("Success")
test/functional/feature_pruning.py:        self.log.info("Success")
test/functional/feature_pruning.py:        self.log.info("Success")
test/functional/feature_pruning.py:        self.log.info("Success")
test/functional/feature_reindex.py:        self.log.info("Success")
test/functional/feature_reindex.py:        self.log.info("Success")
```

stickies-v reviewed (commented):
Concept ACK: reindex validation behaviour being different to IBD validation behaviour is undesirable and should be avoided.

Slight approach NACK: I think this is probably the smallest diff we can get to improve reindex behaviour, but I think there are drawbacks that outweigh the benefits:
1) it still doesn't make IBD and reindex validation behaviour the same: with the current approach blocks that are NOT part of the assumevalid chain will still be validated as assumevalid. While it doesn't seem likely that this can currently be abused by an attacker, I'm not certain and I'd prefer avoiding having to think about it now and in the future.
2) this makes validation slightly more complex by adding another edge case to solve a problem which should only (?) occur in case of corruption, most likely caused by faulty hardware. While I'm not against improving reindex behaviour, I'm not convinced it's important enough to complicate overall validation logic. If the user has fixed their hardware faults, just running a fresh IBD doesn't seem like a big deal for these rare occurences?

Generally speaking, I think we should move towards IBD and reindex validation behaviour converging rather than diverging. Three alternative approaches, sorted by decreasing feasibility, include:
1) Log an error that assumevalid is enabled but reindex could not find the specified assumevalid header, print the highest available header in blockindex and inform the user that they can restart with `-assumevalid=0` or `-assumevalid=<best_available_header>`or a fresh IBD, and then shutdown the node.
2) rearchitect our blockindex storage so that we don't need reindex, or so that it doesn't wipe the assumevalid state. Relevant work is being done in https://github.com/bitcoin/bitcoin/pull/32427, and @cfields shared some relevant thoughts offline too.
3) generalize validation logic so that there is no difference between IBD and reindex, making it possible to do p2p during reindex

My preference for this PR would be to implement approach 1) to not change logic but just increase visibility for the user, and then actually improve things by improving our architecture (approaches 2) and/or 3)) rather than patching things up.

hodlinator commented on src/validation.cpp:
```diff
@@ -2433,13 +2433,22 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
         BlockMap::const_iterator it{m_blockman.m_block_index.find(m_chainman.AssumedValidBlock())};
-        if (it == m_blockman.m_block_index.end()) {
+        bool is_present_in_index{it != m_blockman.m_block_index.end()};
+        bool is_reindexing{m_chainman.m_blockman.IsReindexing()};
```
nit: `const bool` might be nicer, although the scope is small.

hodlinator reviewed (commented):
Agree this issue would be good to fix!

However also agree with https://github.com/bitcoin/bitcoin/pull/31615#pullrequestreview-3403430467 that logging a warning/error and doing some kind of higher level change instead of patching things up here would be preferable.

#### Higher level proposal

Add an earlier phase for `-reindex` where it rebuilds the headers chain/`BlockIndex` from blocks on disk, and if we're still below minimumchainwork, perform headerssync to populate the index prior to this phase of processing blocks.

A condition for exiting this new phase would be that we fulfill minimumchainwork, that way we shouldn't (ever?) have an issue where best header is below minimum chainwork when we reach `ConnectBlock()`.

There may be a concern that some of the headers taken from blocks on disk somehow got corrupted, leading to us filling the block index with garbage. But if we verify headers connect by prev-hash and obey thresholds and difficulty adjustments before feeding them to the index we should be okay (same as we do for headerssync).

#### Alternative

An alternative to the above would be to have a completely different fast path for `-reindex` where if the block exists on disk, we only do minimal hashing to verify that the entire block data is not corrupt. This assumes we do more complete validation before storing blocks to disk (and doesn't account for switching between binaries with different consensus rules).

This would however enable a malicious party to give invalid blocks data to a victim while offline, "here, I optimized IBD for you". But it would still require an eclipse attack / network split to maintain.

#### Sidebar

(There was an issue with the prior 661a32305c13fdffc7dc4420aa44eb503739fc12 where if we halt IBD before reaching the assumevalid block, and restart with -reindex, it looks like we would have disabled script validation even though the block was younger than 2 weeks.

https://github.com/bitcoin/bitcoin/commit/661a32305c13fdffc7dc4420aa44eb503739fc12#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R2462

This has been fixed in newer pushes AFAICS).
